File: 11-container-with-most-water/container-with-most-water.go
func maxArea(height []int) int {
    left, right := 0, len(height)-1
    maxWater := 0

    for left < right {
        // width between lines
        width := right - left

        // height is limited by the shorter line
        h := height[left]
        if height[right] < h {
            h = height[right]
        }

        // calculate area
        area := width * h
        if area > maxWater {
            maxWater = area
        }

        // move the pointer with smaller height
        if height[left] < height[right] {
            left++
        } else {
            right--
        }
    }

    return maxWater
}

File: 11-container-with-most-water/container-with-most-water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:

        i = 0
        j = len(height)-1

        maxWater = 0
        while i < j:
            multfactor = min(height[i], height[j])

            currWater = multfactor*(j-i)
            maxWater = max(maxWater, currWater)
            
            if height[i] > height[j]:
                j-=1
            elif height[i] < height[j]:
                i+=1
            else:
                i+=1
        
        return maxWater